js/background.js

```diff
@@ -228,36 +359,40 @@ function SaveCollection()
 	{
 		timestamp: Date.now(),
 		tabsCount: tabs.length,
-		titles: tabs.map(tab => tab.title ?? ""),
-		links: tabs.map(tab => tab.url ?? ""),
-		icons: tabs.map(tab => tab.favIconUrl ?? ""),
-		thumbnails: tabs.map(tab => thumbnails.find(i => i.tabId == tab.id)?.url ?? "")
+		titles: tabs.map(tab => Truncate(tab.title || "", 100)),
+		links: tabs.map(tab => tab.url ?? "")
 	};
+	tabs.forEach(tab => {//For each tab to save, Add relevant thumbnails in the thumbnails object
+		thumbnails[tab.url]={
+			"pageCapture": sessionCaptures[tab.url] ?? thumbnails[tab.url]?.pageCapture ?? "",
+			"iconUrl":tab.favIconUrl ?? ""
+		}
+	})
 
-	collections.unshift(collection);
-	chrome.storage.sync.set({ "sets": JSON.stringify(collections) })
-
-	var newTabId;
-	chrome.tabs.create({}, (tab) =>
-	{
-		newTabId = tab.id;
-		chrome.tabs.remove(tabsToSave.filter(i => !i.pinned && i.id != newTabId).map(tab => tab.id));
-	});
-
-	UpdateTheme();
+	UpdateStorages({["set_"+collection.timestamp]:collection},()=>
+		chrome.tabs.create({}, (tab) =>
+		{
+			var newTabId = tab.id;
+			chrome.tabs.remove(tabsToSave.filter(i => !i.pinned && i.id != newTabId).map(tab => tab.id));
+		}),
+		(error)=>alert(chrome.i18n.getMessage("errorSavingTabs"))
+	);
 }
 
-function DeleteCollection(collectionIndex)
+function DeleteCollection(collectionKey)
 {
-	collections = collections.filter(i => i != collections[collectionIndex]);
-	chrome.storage.sync.set({ "sets": JSON.stringify(collections) })
-
-	UpdateTheme();
+	var deletedUrls=collections[collectionKey].links
+	delete collections[collectionKey]
+	ForEachUnusedUrl(deletedUrls,(url)=>{
+		delete thumbnails[url]
+		UpdateStorages({})//Updates the thumbnails storage only
```

Comment:
Why do we need to call `UpdateStorages()` after every update to thumbnails array? Why can't we just do it after? `ForEachUnusedUrl` doesn't seem to be async or something

Reply:
You're right, we can do it after. It was initially async while I was working on it, hence the callback mechanism, and I didn't change that. I'll get the update out of the ForEachUnusedUrl - nice catch again, thank you !